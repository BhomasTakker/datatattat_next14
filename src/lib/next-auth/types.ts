import { ObjectId } from "mongoose";
import { GithubProfile } from "next-auth/providers/github";
import { GoogleProfile } from "next-auth/providers/google";
import { SpotifyProfile } from "next-auth/providers/spotify";

export const providers = {
	google: "google",
	github: "github",
	spotify: "spotify",
} as const;

export type Providers = (typeof providers)[keyof typeof providers];
export type ProviderFunction = (
	profile: Profile,
	id: Providers,
	username: string
) => any;

// This prob shouldn't be here
export type User = {
	signup_completed: boolean;
	signin_method: Providers;
	signin_name: string;
	signin_email: string;
	avatar: string;
	username: string;
	role: string;
	_id: ObjectId;
};

export type Profile = GoogleProfile | GithubProfile | SpotifyProfile;
