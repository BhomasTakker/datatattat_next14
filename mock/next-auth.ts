import { Profile } from "@/lib/next-auth/types";
import "@testing-library/jest-dom";

jest.mock("next-auth", () => ({
	getServerSession: jest.fn(),
}));

jest.mock("../src/lib/next-auth/providers/github", () => ({
	GITHUB: {
		id: "github",
		name: "GitHub",
		type: "oauth",
		version: "2.0",
		scope: "read:user user:email",
		profile(profile: Profile) {
			return { id: profile.id, name: profile.name, email: profile.email };
		},
	},
}));

jest.mock("../src/lib/next-auth/providers/google", () => ({
	GOOGLE: {
		id: "google",
		name: "Google",
		type: "oauth",
		version: "2.0",
		scope: "read:user user:email",
		profile(profile: Profile) {
			return { id: profile.id, name: profile.name, email: profile.email };
		},
	},
}));
