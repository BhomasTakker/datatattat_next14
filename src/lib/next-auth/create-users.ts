import { GithubProfile } from "next-auth/providers/github";
import { GoogleProfile } from "next-auth/providers/google";
import { SpotifyProfile } from "next-auth/providers/spotify";
import { Providers } from "./types";

export const createGoogleUser = (
	profile: GoogleProfile,
	id: Providers,
	username: string
) => ({
	signup_completed: false,
	signin_method: id,
	signin_name: profile.name,
	signin_email: profile.email || "", // shouldn't be nothing surely?
	avatar: profile.picture,
	username,
	role: "standard",
});

export const createGithubUser = (
	profile: GithubProfile,
	id: Providers,
	username: string
) => ({
	signup_completed: false,
	signin_method: id,
	signin_name: profile.login,
	signin_email: profile.email || "", // shouldn't be nothing surely?
	avatar: profile.avatar_url,
	username,
	role: "standard",
});

export const createSpotifyUser = (
	profile: SpotifyProfile,
	id: Providers,
	username: string
) => ({
	signup_completed: false,
	signin_method: id,
	signin_name: profile.login,
	signin_email: profile.email || "", // shouldn't be nothing surely?
	avatar: profile.avatar_url,
	username,
	role: "standard",
});
