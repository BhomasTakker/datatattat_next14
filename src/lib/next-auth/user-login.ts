import { checkAndCreateUsername } from "@/actions/signup/check-create-username";
import {
	createNewUser,
	getUserBySignInEmail,
} from "../mongo/actions/user/user";
import { connectToMongoDB } from "../mongo/db";
import { Profile, ProviderFunction, providers, Providers, User } from "./types";
import {
	createGoogleUser,
	createGithubUser,
	createSpotifyUser,
} from "./create-users";

export const providerMap = new Map<Providers, ProviderFunction>([
	[providers.google, createGoogleUser as ProviderFunction],
	[providers.github, createGithubUser as ProviderFunction],
	[providers.spotify, createSpotifyUser as ProviderFunction],
]);

export const createUserObject = (
	profile: Profile,
	id: Providers,
	username: string
) => {
	const createUser = providerMap.get(id);
	if (!createUser) {
		throw new Error("Invalid provider");
	}
	return createUser(profile, id, username);
};

export const loginOrSignUp = async (
	profile: Profile,
	id: Providers
): Promise<User> => {
	// call generic function and pass in profile and an id
	await connectToMongoDB();
	// try catch and reject if fails or no email by email?
	let user = await getUserBySignInEmail(profile.email || "");

	const username = profile.name || profile.login;
	// do we need some kind of fail safe? what if this fails?
	// reject and dont create user
	const uniqueUsername = await checkAndCreateUsername(username);

	if (!user) {
		user = await createNewUser(createUserObject(profile, id, uniqueUsername));
	}

	return user;
};
