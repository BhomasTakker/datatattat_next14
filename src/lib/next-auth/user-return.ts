import { Profile, providers, Providers, User } from "./types";

const githubUsereturn = (profile: Profile, user: User) => {
	return {
		...profile,
		name: user.username,
		image: user.avatar,
		id: profile.id.toString(),
		user_id: user._id.toString(),
	};
};

const googleUserReturn = (profile: Profile, user: User) => {
	console.log("user", user);
	return {
		...profile,
		name: user.username,
		image: user.avatar,
		id: user._id.toString(),
		user_id: user._id.toString(),
	};
};

const providerReturnMap = new Map<Providers, any>([
	[providers.google, googleUserReturn],
	[providers.github, githubUsereturn],
]);

export const returnUser = (profile: Profile, user: User, id: Providers) => {
	const userReturn = providerReturnMap.get(id);
	if (!userReturn) {
		throw new Error("Invalid provider");
	}
	return userReturn(profile, user);
};
