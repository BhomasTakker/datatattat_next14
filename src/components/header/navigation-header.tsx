import { AiFillHome } from "react-icons/ai";
import styles from "./navigation-header.module.scss";
import Link from "next/link";
import { PATHS } from "@/lib/routing/paths";
import { getServerSession } from "next-auth";
import { options } from "@/app/api/auth/[...nextauth]/options";
import { UserMenu } from "./user-menu/user-menu";
import { BiSolidUser } from "react-icons/bi";
import { getUserById } from "@/lib/mongo/actions/user";
import { Session } from "@/types/auth/session";

const Logo = () => {
	return (
		<Link aria-label="Home" href="/">
			<div className={styles.homeContainer}>
				<AiFillHome className={styles.logo} />
			</div>
		</Link>
	);
};

const Datatattat = () => {
	return (
		<div className={styles.logoContainer}>
			<Link href={"/"} key={"Home"} aria-hidden>
				{/* Pull from state - We should have a state config */}
				<h2 className={styles.datatattat}>DATATATTAT</h2>
			</Link>
		</div>
	);
};

const SignInButton = () => {
	return (
		// callback to current page
		<Link className={styles.simpleButton} href={PATHS.signIn("/")}>
			<BiSolidUser className={styles.logo} />
			<p>Login</p>
		</Link>
	);
};

export const NavigationHeader = async () => {
	const session = (await getServerSession(options)) as Session;

	let menuComponent = <div />;

	// Warning
	// Here may look like a bug
	// If we have an active session in local storage i.e. for Development
	// If we then are running in 'Production' we will not have a valid session
	// So the user button to sign in or modify will not be shown
	if (!session) {
		menuComponent = <SignInButton />;
	} else {
		// We can have in local storage the wrong user - from Prod or Staging etc
		const { user } = session;
		const { user_id } = user;
		const userData = await getUserById(user_id);

		if (userData) {
			const { username, avatar } = userData;
			menuComponent = <UserMenu username={username} avatar={avatar} />;
		}
	}

	return (
		<nav className={styles.root}>
			<Logo />
			<Datatattat />
			{menuComponent}
		</nav>
	);
};
