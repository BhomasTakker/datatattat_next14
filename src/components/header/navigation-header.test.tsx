import "@testing-library/jest-dom";
import { render, screen, waitFor } from "@testing-library/react";
import { NavigationHeader } from "./navigation-header";
import { getServerSession } from "next-auth";

jest.mock("./user-menu/user-menu", () => {
	return {
		UserMenu: () => {
			return <div>User Menu</div>;
		},
	};
});

jest.mock("next-auth", () => {
	return {
		getServerSession: jest.fn().mockResolvedValue({
			user: {
				user_id: "user_id",
			},
		}),
	};
});

jest.mock("../../app/api/auth/[...nextauth]/options", () => {
	return {
		options: {},
	};
});
// copilot did this!
jest.mock("../../lib/mongo/actions/user/user", () => {
	return {
		getUserById: jest.fn().mockResolvedValue({
			role: "role",
			username: "username",
			avatar: "avatar",
		}),
	};
});

describe("Navigation Header Test Suite", () => {
	it("renders a nav element", async () => {
		render(await NavigationHeader());
		await waitFor(() => {
			const nav = screen.getByRole("navigation");
			expect(nav).toBeInTheDocument();
		});
	});

	it("renders a logo", async () => {
		render(await NavigationHeader());
		await waitFor(() => {
			const logo = screen.getByLabelText("Home");
			expect(logo).toBeInTheDocument();
		});
	});

	it("renders Site Name", async () => {
		render(await NavigationHeader());
		await waitFor(() => {
			const datatattat = screen.getByText("DATATATTAT");
			expect(datatattat).toBeInTheDocument();
		});
	});

	it("renders User Menu when signed in", async () => {
		render(await NavigationHeader());
		await waitFor(() => {
			const userMenu = screen.getByText("User Menu");
			expect(userMenu).toBeInTheDocument();
		});
	});

	it("renders Sign In Button when not signed in", async () => {
		(getServerSession as jest.Mock).mockResolvedValueOnce(false);
		render(await NavigationHeader());
		await waitFor(() => {
			const signInButton = screen.getByTestId("sign-in");
			expect(signInButton).toBeInTheDocument();
		});
	});

	it("renders signed in navigation header unchanged", async () => {
		const { container } = render(await NavigationHeader());
		expect(container).toMatchSnapshot();
	});

	it("renders signed out navigation header unchanged", async () => {
		(getServerSession as jest.Mock).mockResolvedValueOnce(false);
		const { container } = render(await NavigationHeader());

		expect(container).toMatchSnapshot();
	});
});
