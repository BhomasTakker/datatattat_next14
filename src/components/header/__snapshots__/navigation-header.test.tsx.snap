// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation Header Test Suite renders signed in navigation header unchanged 1`] = `
<div>
  <nav
    class="root"
  >
    <a
      aria-label="Home"
      href="/"
    >
      <div
        class="homeContainer"
      >
        <svg
          class="logo"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 1024 1024"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M946.5 505L534.6 93.4a31.93 31.93 0 0 0-45.2 0L77.5 505c-12 12-18.8 28.3-18.8 45.3 0 35.3 28.7 64 64 64h43.4V908c0 17.7 14.3 32 32 32H448V716h112v224h265.9c17.7 0 32-14.3 32-32V614.3h43.4c17 0 33.3-6.7 45.3-18.8 24.9-25 24.9-65.5-.1-90.5z"
          />
        </svg>
      </div>
    </a>
    <div
      class="logoContainer"
    >
      <a
        aria-hidden="true"
        href="/"
      >
        <h2
          class="datatattat"
        >
          DATATATTAT
        </h2>
      </a>
    </div>
    <div>
      User Menu
    </div>
  </nav>
</div>
`;

exports[`Navigation Header Test Suite renders signed out navigation header unchanged 1`] = `
<div>
  <nav
    class="root"
  >
    <a
      aria-label="Home"
      href="/"
    >
      <div
        class="homeContainer"
      >
        <svg
          class="logo"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 1024 1024"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M946.5 505L534.6 93.4a31.93 31.93 0 0 0-45.2 0L77.5 505c-12 12-18.8 28.3-18.8 45.3 0 35.3 28.7 64 64 64h43.4V908c0 17.7 14.3 32 32 32H448V716h112v224h265.9c17.7 0 32-14.3 32-32V614.3h43.4c17 0 33.3-6.7 45.3-18.8 24.9-25 24.9-65.5-.1-90.5z"
          />
        </svg>
      </div>
    </a>
    <div
      class="logoContainer"
    >
      <a
        aria-hidden="true"
        href="/"
      >
        <h2
          class="datatattat"
        >
          DATATATTAT
        </h2>
      </a>
    </div>
    <a
      class="simpleButton"
      href="/api/auth/signin?callbackUrl=/"
    >
      <svg
        class="logo"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M7.5 6.5C7.5 8.981 9.519 11 12 11s4.5-2.019 4.5-4.5S14.481 2 12 2 7.5 4.019 7.5 6.5zM20 21h1v-1c0-3.859-3.141-7-7-7h-4c-3.86 0-7 3.141-7 7v1h17z"
        />
      </svg>
      <p>
        Login
      </p>
    </a>
  </nav>
</div>
`;
